File: Samples/CustomAnimation/C++/Src/ElasticAnimation.cpp
////////////////////////////////////////////////////////////////////////////////////////////////////
// NoesisGUI - http://www.noesisengine.com
// Copyright (c) 2013 Noesis Technologies S.L. All Rights Reserved.
////////////////////////////////////////////////////////////////////////////////////////////////////


#include "ElasticAnimation.h"

#include <NsCore/ReflectionImplement.h>
#include <NsCore/ReflectionImplementEnum.h>
#include <NsCore/Nullable.h>
#include <NsGui/UIElementData.h>
#include <NsGui/PropertyMetadata.h>
#include <NsGui/AnimationClock.h>


using namespace CustomAnimation;
using namespace Noesis;


////////////////////////////////////////////////////////////////////////////////////////////////////
static float EaseOut(float t, float start, float delta, float springiness, float oscillations)
{
    float val = powf(1.0f - t, springiness) * cosf(TwoPi * t * oscillations);
    val = delta - val * delta;
    val += start;
    return val;
}

////////////////////////////////////////////////////////////////////////////////////////////////////
static float EaseIn(float t, float start, float delta, float springiness, float oscillations)
{
    float val = powf(t, springiness) * cosf(TwoPi * t * oscillations);
    val = val * delta;
    val += start;
    return val;
}

////////////////////////////////////////////////////////////////////////////////////////////////////
static float EaseInOut(float t, float start, float delta, float springiness, float oscillations)
{
    float val;

    if (t <= 0.5f)
    {
        val = EaseIn(t * 2.0f, start, delta * 0.5f, springiness, oscillations);
    }
    else
    {
        val = EaseOut((t - 0.5f) * 2.0f, start, delta * 0.5f, springiness, oscillations);
        val += delta * 0.5f;
    }

    return val;
}

////////////////////////////////////////////////////////////////////////////////////////////////////
ElasticAnimation::EdgeBehaviorEnum ElasticAnimation::GetEdgeBehavior() const
{
    return GetValue<EdgeBehaviorEnum>(EdgeBehaviorProperty);
}

////////////////////////////////////////////////////////////////////////////////////////////////////
void ElasticAnimation::SetEdgeBehavior(EdgeBehaviorEnum value)
{
    SetValue<EdgeBehaviorEnum>(EdgeBehaviorProperty, value);
}

////////////////////////////////////////////////////////////////////////////////////////////////////
float ElasticAnimation::GetSpringiness() const
{
    return GetValue<float>(SpringinessProperty);
}

////////////////////////////////////////////////////////////////////////////////////////////////////
void ElasticAnimation::SetSpringiness(float value)
{
    SetValue<float>(SpringinessProperty, value);
}

////////////////////////////////////////////////////////////////////////////////////////////////////
float ElasticAnimation::GetOscillations() const
{
    return GetValue<float>(OscillationsProperty);
}

////////////////////////////////////////////////////////////////////////////////////////////////////
void ElasticAnimation::SetOscillations(float value)
{
    SetValue<float>(OscillationsProperty, value);
}

////////////////////////////////////////////////////////////////////////////////////////////////////
Nullable<float> ElasticAnimation::GetFrom() const
{
    return GetValue<Nullable<float>>(FromProperty);
}

////////////////////////////////////////////////////////////////////////////////////////////////////
void ElasticAnimation::SetFrom(Nullable<float> value)
{
    SetValue<Nullable<float>>(FromProperty, value);
}

////////////////////////////////////////////////////////////////////////////////////////////////////
Nullable<float> ElasticAnimation::GetTo() const
{
    return GetValue<Nullable<float>>(ToProperty);
}

////////////////////////////////////////////////////////////////////////////////////////////////////
void ElasticAnimation::SetTo(Nullable<float> value)
{
    SetValue<Nullable<float>>(ToProperty, value);
}

////////////////////////////////////////////////////////////////////////////////////////////////////
float ElasticAnimation::GetCurrentValueCore(float defaultOrigin, float, AnimationClock* clock)
{
    float start = GetFrom().HasValue() ? GetFrom().GetValue() : defaultOrigin;
    float delta = GetTo().HasValue() ? GetTo().GetValue() - start : defaultOrigin - start;
    float springiness = GetSpringiness();
    float oscillations = GetOscillations();

    switch (GetEdgeBehavior())
    {
        case EdgeBehaviorEnum::EaseIn:
        {
            return EaseIn(clock->GetCurrentProgress(), start, delta, springiness, oscillations);
        }
        case EdgeBehaviorEnum::EaseOut:
        {
            return EaseOut(clock->GetCurrentProgress(), start, delta, springiness, oscillations);
        }
        case EdgeBehaviorEnum::EaseInOut:
        {
            return EaseInOut(clock->GetCurrentProgress(), start, delta, springiness, oscillations);
        }
        default:
            NS_ASSERT_UNREACHABLE;
    }
}

////////////////////////////////////////////////////////////////////////////////////////////////////
Ptr<Freezable> ElasticAnimation::CreateInstanceCore() const
{
    return *new ElasticAnimation();
}

////////////////////////////////////////////////////////////////////////////////////////////////////
NS_BEGIN_COLD_REGION

NS_IMPLEMENT_REFLECTION_ENUM(CustomAnimation::ElasticAnimation::EdgeBehaviorEnum)
{
    NsVal("EaseIn", ElasticAnimation::EdgeBehaviorEnum::EaseIn);
    NsVal("EaseOut", ElasticAnimation::EdgeBehaviorEnum::EaseOut);
    NsVal("EaseInOut", ElasticAnimation::EdgeBehaviorEnum::EaseInOut);
}

NS_IMPLEMENT_REFLECTION(CustomAnimation::ElasticAnimation, "CustomAnimation.ElasticAnimation")
{
    UIElementData* data = NsMeta<UIElementData>(TypeOf<SelfClass>());

    data->RegisterProperty<ElasticAnimation::EdgeBehaviorEnum>(EdgeBehaviorProperty, "EdgeBehavior",
        PropertyMetadata::Create(ElasticAnimation::EdgeBehaviorEnum::EaseIn));

    data->RegisterProperty<float>(SpringinessProperty, "Springiness", PropertyMetadata::Create(3.0f));
    data->RegisterProperty<float>(OscillationsProperty, "Oscillations", PropertyMetadata::Create(10.0f));

    data->RegisterProperty<Nullable<float>>(FromProperty, "From", PropertyMetadata::Create(Nullable<float>()));
    data->RegisterProperty<Nullable<float>>(ToProperty, "To", PropertyMetadata::Create(Nullable<float>()));
}

NS_END_COLD_REGION

////////////////////////////////////////////////////////////////////////////////////////////////////
const DependencyProperty* ElasticAnimation::EdgeBehaviorProperty;
const DependencyProperty* ElasticAnimation::SpringinessProperty;
const DependencyProperty* ElasticAnimation::OscillationsProperty;
const DependencyProperty* ElasticAnimation::FromProperty;
const DependencyProperty* ElasticAnimation::ToProperty;

File: Samples/CustomAnimation/Unity/Assets/Samples/NoesisGUI/CustomAnimation/ElasticAnimation.cs
// Copyright (c) Microsoft. All rights reserved.
// Licensed under the MIT license. See LICENSE file in the project root for full license information.

#if UNITY_5_3_OR_NEWER
#define NOESIS
using Noesis;
using System;
using Float = System.Single;
#else
using System;
using System.Windows;
using System.Windows.Media.Animation;
using Float = System.Double;
#endif

namespace CustomAnimation
{
    /// <summary>
    ///     ElasticAnimation - like something attached to a rubber band
    /// </summary>
    public class ElasticAnimation : DoubleAnimationBase
    {
        public enum EdgeBehaviorEnum
        {
            EaseIn,
            EaseOut,
            EaseInOut
        }

        public static readonly DependencyProperty EdgeBehaviorProperty = DependencyProperty.Register(
            "EdgeBehavior", typeof(EdgeBehaviorEnum), typeof(ElasticAnimation),
            new PropertyMetadata(EdgeBehaviorEnum.EaseIn));

        public static readonly DependencyProperty SpringinessProperty = DependencyProperty.Register(
            "Springiness", typeof(Float), typeof(ElasticAnimation),
            new PropertyMetadata((Float)3.0));

        public static readonly DependencyProperty OscillationsProperty = DependencyProperty.Register(
            "Oscillations", typeof(Float), typeof(ElasticAnimation),
            new PropertyMetadata((Float)10.0));

        public static readonly DependencyProperty FromProperty = DependencyProperty.Register(
            "From", typeof(Float?), typeof(ElasticAnimation),
            new PropertyMetadata(null));

        public static readonly DependencyProperty ToProperty = DependencyProperty.Register(
            "To", typeof(Float?), typeof(ElasticAnimation),
            new PropertyMetadata(null));

        /// <summary>
        ///     which side gets the effect
        /// </summary>
        public EdgeBehaviorEnum EdgeBehavior
        {
            get { return (EdgeBehaviorEnum) GetValue(EdgeBehaviorProperty); }
            set { SetValue(EdgeBehaviorProperty, value); }
        }

        /// <summary>
        ///     how much springiness is there in the effect
        /// </summary>
        public Float Springiness
        {
            get { return (Float) GetValue(SpringinessProperty); }
            set { SetValue(SpringinessProperty, value); }
        }

        /// <summary>
        ///     number of oscillations in the effect
        /// </summary>
        public Float Oscillations
        {
            get { return (Float) GetValue(OscillationsProperty); }
            set { SetValue(OscillationsProperty, value); }
        }

        /// <summary>
        ///     Specifies the starting value of the animation.
        /// </summary>
        public Float? From
        {
            get { return (Float?) GetValue(FromProperty); }
            set { SetValue(FromProperty, value); }
        }

        /// <summary>
        ///     Specifies the ending value of the animation.
        /// </summary>
        public Float? To
        {
            get { return (Float?) GetValue(ToProperty); }
            set { SetValue(ToProperty, value); }
        }

        protected override Float GetCurrentValueCore(Float defaultOriginValue, Float defaultDestinationValue,
            AnimationClock clock)
        {
            Float returnValue;
            var start = From ?? defaultOriginValue;
            var delta = To - start ?? defaultOriginValue - start;
            switch (EdgeBehavior)
            {
                case EdgeBehaviorEnum.EaseIn:
                    returnValue = EaseIn(clock.CurrentProgress.Value, start, delta, Springiness, Oscillations);
                    break;
                case EdgeBehaviorEnum.EaseOut:
                    returnValue = EaseOut(clock.CurrentProgress.Value, start, delta, Springiness, Oscillations);
                    break;
                default:
                    returnValue = EaseInOut(clock.CurrentProgress.Value, start, delta, Springiness, Oscillations);
                    break;
            }
            return returnValue;
        }

#if !NOESIS
        protected override Freezable CreateInstanceCore() => new ElasticAnimation();
#endif

        private static Float EaseOut(Float timeFraction, Float start, Float delta, Float springiness,
            Float oscillations)
        {
            // math magic: The cosine gives us the right wave, the timeFraction * the # of oscillations is the 
            // frequency of the wave, and the amplitude (the exponent) makes the wave get smaller at the end
            // by the "springiness" factor. This is extremely similar to the bounce equation.
            var returnValue = Math.Pow((1 - timeFraction), springiness)
                              *Math.Cos(2*Math.PI*timeFraction*oscillations);
            returnValue = delta - (returnValue*delta);
            returnValue += start;
            return (Float)returnValue;
        }

        private static Float EaseIn(Float timeFraction, Float start, Float delta, Float springiness,
            Float oscillations)
        {
            // math magic: The cosine gives us the right wave, the timeFraction * the # of oscillations is the 
            // frequency of the wave, and the amplitude (the exponent) makes the wave get smaller at the beginning
            // by the "springiness" factor. This is extremely similar to the bounce equation. 
            var returnValue = Math.Pow((timeFraction), springiness)
                              *Math.Cos(2*Math.PI*timeFraction*oscillations);
            returnValue = returnValue*delta;
            returnValue += start;
            return (Float)returnValue;
        }

        private static Float EaseInOut(Float timeFraction, Float start, Float delta, Float springiness,
            Float oscillations)
        {
            Float returnValue;

            // we cut each effect in half by multiplying the time fraction by two and halving the distance.
            if (timeFraction <= 0.5)
            {
                return EaseIn(timeFraction*2, start, delta/2, springiness, oscillations);
            }
            returnValue = EaseOut((timeFraction - 0.5f)*2, start, delta/2, springiness, oscillations);
            returnValue += (delta/2);
            return (Float)returnValue;
        }
    }
}